napari/tests/test_viewer.py

```diff
@@ -28,6 +28,12 @@ def test_viewer(qtbot):
     # Run all class keybindings
     for func in viewer.class_keymap.values():
         func(viewer)
+        # the `play` keybinding calls QtDims.play_dim(), which then creates a new QThread.
+        # we must then run the keybinding a second time, which will call QtDims.stop(),
+        # otherwise the thread will be killed at the end of the test without cleanup,
+        # causing a segmentation fault.  (though the tests still pass)
+        if func.__name__ == 'play':
+            func(viewer)
```

Comment:
This feels kinda messy and brittle... Could there perhaps be a context manager that cleanly deletes running threads? I can imagine more and more keybindings needing special treatment like this, which would be a nightmare without a unified way to handle them...

But, I'm reasonably ok with this workaround. Just want to make sure we've explored all the options.

Reply:
looked into it for a while, and while I agree this is ugly, I didn't like the alternatives I found.
The only way I could figure out to "find all QThreads" is to start at some high level object (like `QtWidgets.QApplication.instance().topLevelWidgets()`) and recurse through all `children()` looking for instances of `QThreads` to kill.  That might work, but may also eventually lead to unintended side effects of its own if we go killing all threads with reckless abandon.  And if we only try to kill certain threads (like `AnimationThread`), we're right back to checking for something specific (like `func.__name__ == 'play'`).  I agree it seems like a handy trick to have though, so let me know if you'd like me to push  harder on it.